File: client-android/serenada-core/src/test/java/app/serenada/core/fakes/RegistryTestHarness.kt
package app.serenada.core.fakes

import app.serenada.core.ForegroundMediaArbiter
import app.serenada.core.JoinAndSwitchResult
import app.serenada.core.JoinResult
import app.serenada.core.RoomRef
import app.serenada.core.SerenadaCallRegistry
import app.serenada.core.SerenadaConfig
import app.serenada.core.SerenadaSession
import app.serenada.core.SwitchResult
import app.serenada.core.call.CallMediaRole
import app.serenada.core.call.SerenadaAudioCoordinator
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import okhttp3.OkHttpClient
import org.robolectric.RuntimeEnvironment
import org.robolectric.Shadows
import org.robolectric.shadows.ShadowLooper
import org.webrtc.PeerConnection

/**
 * One fake-backed managed session created by the registry under test. Bundles the
 * session with its fakes so a test can assert per-call media/coordinator behavior.
 */
internal class FakeManagedSession(
    val roomId: String,
    val coordinator: CountingTestCoordinator,
    val fakeProvider: FakeSignalingProvider,
    val fakeAudio: FakeAudioController,
    val fakeMedia: FakeMediaEngine,
    val session: SerenadaSession,
) {
    /**
     * Drive this held session's signaling to settle (Joining -> Waiting/InCall) so
     * the registry's bounded held-join await resolves. [peers] is the number of
     * participants reported in the joined event (>=2 => InCall).
     */
    fun settle(localCid: String = "cid-$roomId", peers: Int = 1) {
        fakeProvider.simulateConnected("ws")
        ShadowLooper.idleMainLooper()
        val participants = buildList {
            add(localCid to 1L)
            for (i in 1 until peers) add("remote-$roomId-$i" to (i + 1).toLong())
        }
        fakeProvider.simulateJoined(
            peerId = localCid,
            participants = participants,
            hostPeerId = localCid,
        )
        ShadowLooper.idleMainLooper()
    }

    /** Force the held room join to fail (the registry returns JoinFailed). */
    fun fail(code: String = "join_failed", message: String = "boom") {
        fakeProvider.simulateError(code, message)
        ShadowLooper.idleMainLooper()
    }
}

/**
 * Test harness for [SerenadaCallRegistry]. Supplies the registry's internal
 * session-factory seam with fake-backed sessions so each created managed call is
 * fully observable (per-call coordinator/media/provider) without real WebRTC,
 * permissions, or network. Each registry gets a DEDICATED process-global arbiter
 * (reset on construction) so a lease held by a prior sequential test cannot leak
 * in (contract §2 / Phase 2 green-gate note).
 */
internal class RegistryTestHarness(
    private val grantPermissions: Boolean = true,
    private val defaultVideoEnabled: Boolean = false,
    /**
     * When set, each managed session's signaling channel is vended by THIS single
     * global v2 service through the REAL [app.serenada.core.SerenadaCore] seam
     * ([app.serenada.core.SerenadaCore.createSignalingProvider] →
     * [FakeMultiSessionSignalingProvider.openSession]) instead of a standalone
     * [FakeSignalingProvider]. Drives F2 per-session channel isolation.
     */
    private val multiSessionProvider: FakeMultiSessionSignalingProvider? = null,
) {
    /** Created sessions, in creation order, keyed by canonical room id. */
    val created = LinkedHashMap<String, FakeManagedSession>()

    /**
     * Real core used ONLY to exercise the multi-session signaling seam when
     * [multiSessionProvider] is set. Sessions still receive fake media/audio (real
     * WebRTC cannot init under Robolectric); only the signaling channel flows through
     * the real seam.
     */
    private val seamCore: app.serenada.core.SerenadaCore? = multiSessionProvider?.let {
        app.serenada.core.SerenadaCore(
            config = SerenadaConfig(multiSessionSignalingProvider = it),
            context = RuntimeEnvironment.getApplication(),
        )
    }

    init {
        ForegroundMediaArbiter.resetForTests()
        if (grantPermissions) {
            Shadows.shadowOf(RuntimeEnvironment.getApplication()).grantPermissions(
                android.Manifest.permission.CAMERA,
                android.Manifest.permission.RECORD_AUDIO,
            )
        }
    }

    /**
     * When set, the NEXT session-factory call THROWS this (and clears itself) instead
     * of building a session. Models a session construction failure (e.g. an
     * unconfigured core) so a registry test can drive the mode-claim-leak guard in
     * `registerOrDedup` (D-native-3).
     */
    var throwOnNextSessionCreate: Throwable? = null

    val registry: SerenadaCallRegistry = SerenadaCallRegistry(
        core = null,
        sessionFactory = { room, role ->
            throwOnNextSessionCreate?.let {
                throwOnNextSessionCreate = null
                throw it
            }
            buildSession(room, role)
        },
    )

    // Same Robolectric main looper the registry/session scopes use.
    private val opScope = CoroutineScope(Dispatchers.Main.immediate)

    /**
     * Idle the main looper until [deferred] completes, then return its value.
     * Advances virtual time in steps so a registry `withTimeout` (release/activate
     * drain) fires deterministically when a test injects a hung release/activation.
     */
    private fun <T> drain(deferred: Deferred<T>): T {
        var guard = 0
        while (!deferred.isCompleted && guard < 200) {
            ShadowLooper.idleMainLooper()
            if (!deferred.isCompleted) {
                // Advance past the longest registry timeout (activate, 12s) so any
                // pending withTimeout cancels.
                Shadows.shadowOf(android.os.Looper.getMainLooper())
                    .idleFor(java.time.Duration.ofMillis(13_000L))
            }
            guard++
        }
        check(deferred.isCompleted) { "registry op did not complete" }
        @OptIn(kotlinx.coroutines.ExperimentalCoroutinesApi::class)
        return deferred.getCompleted()
    }

    /**
     * Run [SerenadaCallRegistry.joinHeld]: launch the op, idle so part A creates +
     * starts the held session, settle that session's signaling so the bounded
     * held-join await resolves, then drain to the result. The newly created
     * session is settled with [peers] participants.
     */
    fun joinHeld(room: RoomRef, peers: Int = 1): JoinResult {
        val before = created.keys.toSet()
        val deferred = opScope.async { registry.joinHeld(room) }
        ShadowLooper.idleMainLooper()
        settleNewlyCreated(before, peers)
        return drain(deferred)
    }

    /** Run [SerenadaCallRegistry.joinAndSwitch] with the same settle orchestration. */
    fun joinAndSwitch(room: RoomRef, peers: Int = 1): JoinAndSwitchResult {
        val before = created.keys.toSet()
        val deferred = opScope.async { registry.joinAndSwitch(room) }
        ShadowLooper.idleMainLooper()
        settleNewlyCreated(before, peers)
        return drain(deferred)
    }

    /**
     * Run [SerenadaCallRegistry.joinHeld] but make the held room join FAIL instead
     * of settling (the session is driven to Error). Returns the [JoinResult].
     */
    fun joinHeldFailing(room: RoomRef): JoinResult {
        val before = created.keys.toSet()
        val deferred = opScope.async { registry.joinHeld(room) }
        ShadowLooper.idleMainLooper()
        newlyCreated(before)?.fail()
        return drain(deferred)
    }

    /** Run [SerenadaCallRegistry.joinAndSwitch] but make the held room join FAIL. */
    fun joinAndSwitchFailing(room: RoomRef): JoinAndSwitchResult {
        val before = created.keys.toSet()
        val deferred = opScope.async { registry.joinAndSwitch(room) }
        ShadowLooper.idleMainLooper()
        newlyCreated(before)?.fail()
        return drain(deferred)
    }

    fun switchToCall(callId: String): SwitchResult {
        val deferred = opScope.async { registry.switchToCall(callId) }
        return drain(deferred)
    }

    fun holdCall(callId: String) {
        drain(opScope.async { registry.holdCall(callId) })
    }

    fun leaveCall(callId: String) {
        drain(opScope.async { registry.leaveCall(callId) })
    }

    fun endCall(callId: String) {
        drain(opScope.async { registry.endCall(callId) })
    }

    fun dismissCall(callId: String) {
        drain(opScope.async { registry.dismissCall(callId) })
    }

    /**
     * Drive [roomKey]'s session to a TERMINAL phase on its OWN (room_ended / remote
     * end) — NOT via a registry leave/end. Idles the main looper so the session's
     * cleanup AND the registry's collector terminal op both run.
     */
    fun simulateRoomEnded(roomKey: String) {
        created[roomKey]!!.fakeProvider.simulateRoomEnded()
        ShadowLooper.idleMainLooper()
    }

    /**
     * Run an arbitrary suspend [block] on the same main-immediate scope the
     * registry/sessions use, draining the looper to completion. Lets a test poke a
     * session's internal hold/resume primitive (e.g.
     * [SerenadaSession.applyHeldRoleInternal]) to model a session-side partial
     * release while the registry still holds the lease token.
     */
    fun <T> runOnMain(block: suspend () -> T): T = drain(opScope.async { block() })

    /**
     * Launch [block] WITHOUT draining, for tests that interleave concurrent
     * registry ops (e.g. a switch racing a still-pending held join). Idle the
     * looper manually between steps; finish with [await]. Unlike [drain], plain
     * [ShadowLooper.idleMainLooper] does not advance virtual time, so a pending
     * held join does NOT hit its timeout while the test interleaves.
     */
    fun <T> launchOp(block: suspend () -> T): Deferred<T> = opScope.async { block() }

    /** Drain [deferred] to completion (advancing virtual time past timeouts if needed). */
    fun <T> await(deferred: Deferred<T>): T = drain(deferred)

    /**
     * Idle the main looper (advancing virtual time past timeouts if needed) until
     * [deferred] completes, WITHOUT unwrapping its result. Unlike [drain]/[await]
     * this tolerates a CANCELLED deferred (whose [getCompleted] would rethrow), so
     * a test can cancel an in-flight op and then let its non-cancellable cleanup run.
     */
    fun idleUntilComplete(deferred: Deferred<*>) {
        var guard = 0
        while (!deferred.isCompleted && guard < 200) {
            ShadowLooper.idleMainLooper()
            if (!deferred.isCompleted) {
                Shadows.shadowOf(android.os.Looper.getMainLooper())
                    .idleFor(java.time.Duration.ofMillis(13_000L))
            }
            guard++
        }
        check(deferred.isCompleted) { "registry op did not complete" }
    }

    private fun newlyCreated(before: Set<String>): FakeManagedSession? {
        val newKey = created.keys.firstOrNull { it !in before } ?: return null
        return created[newKey]
    }

    private fun settleNewlyCreated(before: Set<String>, peers: Int) {
        newlyCreated(before)?.settle(peers = peers)
    }

    private fun roomKeyFor(room: RoomRef): String = when (room) {
        is RoomRef.Id -> room.roomId
        is RoomRef.Url -> room.url
    }

    private fun roomIdFor(room: RoomRef): String = when (room) {
        is RoomRef.Id -> room.roomId
        is RoomRef.Url -> room.url.trimEnd('/').substringAfterLast('/')
    }

    private fun buildSession(room: RoomRef, role: CallMediaRole): SerenadaSession {
        val roomId = roomIdFor(room)
        val coordinator = CountingTestCoordinator()
        // In v2 mode the channel comes from the REAL core seam (openSession(roomId));
        // otherwise a standalone single-session fake (the pre-existing path).
        val fakeProvider: FakeSignalingProvider = if (seamCore != null) {
            seamCore.createSignalingProvider(
                SerenadaConfig(multiSessionSignalingProvider = multiSessionProvider),
                roomId,
            ) as FakeSignalingProvider
        } else {
            FakeSignalingProvider()
        }
        val fakeAudio = FakeAudioController()
        val fakeMedia = FakeMediaEngine()
        fakeProvider.enqueueIceServers(
            Result.success(
                listOf(
                    PeerConnection.IceServer.builder("turn:turn.example.com:3478")
                        .setUsername("user")
                        .setPassword("pass")
                        .createIceServer(),
                ),
            ),
        )
        val session = SerenadaSession(
            roomId = roomId,
            roomUrl = null,
            config = SerenadaConfig(
                signalingProvider = fakeProvider,
                defaultVideoEnabled = defaultVideoEnabled,
                videoMediaEnabled = true,
                audioCoordinator = coordinator,
            ),
            context = RuntimeEnvironment.getApplication(),
            delegate = null,
            okHttpClient = OkHttpClient(),
            initialSignalingProvider = fakeProvider,
            audioController = fakeAudio,
            mediaEngine = fakeMedia,
            clock = FakeSessionClock(),
            initialMediaRole = role,
            // The registry owns the lease; the session must NOT self-acquire.
            acquireForegroundLease = false,
        )
        // Mirror SerenadaCore.joinInternal: start the session immediately.
        session.start()
        created[roomKeyFor(room)] = FakeManagedSession(
            roomId = roomId,
            coordinator = coordinator,
            fakeProvider = fakeProvider,
            fakeAudio = fakeAudio,
            fakeMedia = fakeMedia,
            session = session,
        )
        return session
    }

    fun close() {
        drain(opScope.async { runCatching { registry.close() } })
    }

    fun tearDown() {
        runCatching { close() }
        ShadowLooper.idleMainLooper()
        ForegroundMediaArbiter.resetForTests()
    }
}

/** Audio coordinator that counts activate/deactivate calls; never suspends. */
internal class CountingTestCoordinator : SerenadaAudioCoordinator {
    var activateCalls = 0
        private set
    var deactivateCalls = 0
        private set

    /**
     * When true, the NEXT [activateCallSession] THROWS instead of activating (and
     * clears itself). Models a real audio-coordinator activation failure so a
     * registry test can drive the FIX-A rollback path through the actual
     * coordinator boundary (distinct from
     * [SerenadaSession.failNextForegroundActivationForTest], which short-circuits
     * before the coordinator runs).
     */
    var failNextActivation = false

    /** Append-only log of "activate"/"deactivate" to assert ORDERING across calls. */
    val ops = mutableListOf<String>()

    override suspend fun activateCallSession(intent: app.serenada.core.call.AudioIntent) {
        if (failNextActivation) {
            failNextActivation = false
            throw IllegalStateException("test-injected audio coordinator activation failure")
        }
        activateCalls += 1
        ops += "activate"
    }

    override suspend fun deactivateCallSession() {
        deactivateCalls += 1
        ops += "deactivate"
    }

    override suspend fun applyRouting(device: app.serenada.core.call.AudioDevice) {}
    override suspend fun setMicMuted(muted: Boolean) {}

    override val availableDevices =
        kotlinx.coroutines.flow.MutableStateFlow<List<app.serenada.core.call.AudioDevice>>(emptyList())
    override val effectiveInputDevice =
        kotlinx.coroutines.flow.MutableStateFlow<app.serenada.core.call.AudioDevice?>(null)
    override val effectiveOutputDevice =
        kotlinx.coroutines.flow.MutableStateFlow<app.serenada.core.call.AudioDevice?>(null)
    override val events =
        kotlinx.coroutines.flow.MutableSharedFlow<app.serenada.core.call.AudioCoordinatorEvent>()
}

File: client-ios/SerenadaCore/Tests/SerenadaCoreTests/Helpers/RegistryTestHarness.swift
import Combine
import Foundation
@testable import SerenadaCore

// MARK: - Shared ordered event log

/// Ordered, cross-coordinator event log so a switch test can assert
/// "deactivate-old strictly before activate-new" across TWO different sessions'
/// coordinators (the per-coordinator count-only `FakeAudioCoordinator` cannot
/// express ordering between distinct instances).
@MainActor
final class EventLog {
    private(set) var events: [String] = []
    func append(_ event: String) { events.append(event) }
    func clear() { events.removeAll() }
}

// MARK: - Recording audio coordinator

/// Audio coordinator that records tagged activate/deactivate events into a shared
/// ``EventLog`` so registry tests can assert ordering across sessions. Activation
/// is immediate (no block); for blocking variants see `GatedAudioCoordinator`
/// (activation) and `BlockingDeactivateCoordinator` (deactivation).
@MainActor
final class RecordingAudioCoordinator: SerenadaAudioCoordinator, @unchecked Sendable {
    private let tag: String
    private let log: EventLog
    private(set) var activateEvents = 0
    private(set) var deactivateEvents = 0

    init(tag: String, log: EventLog) {
        self.tag = tag
        self.log = log
    }

    func activateCallSession(intent: AudioIntent) async throws {
        activateEvents += 1
        log.append("\(tag):activate")
    }

    func deactivateCallSession() async {
        deactivateEvents += 1
        log.append("\(tag):deactivate")
    }

    func applyRouting(_ device: AudioDevice) async throws {}
    func setMicMuted(_ muted: Bool) async throws {}

    var availableDevices: AsyncStream<[AudioDevice]> { AsyncStream { _ in } }
    var effectiveInputDevice: AsyncStream<AudioDevice?> { AsyncStream { _ in } }
    var effectiveOutputDevice: AsyncStream<AudioDevice?> { AsyncStream { _ in } }
    var events: AsyncStream<AudioCoordinatorEvent> { AsyncStream { _ in } }
}

/// Audio coordinator whose `deactivateCallSession` can be PAUSED so a test can
/// drive the registry's old-release timeout (the old call's drain never confirms
/// fully-held). Activation is immediate.
@MainActor
final class BlockingDeactivateCoordinator: SerenadaAudioCoordinator, @unchecked Sendable {
    var blockNextDeactivation = false
    private(set) var deactivationInFlight = false
    private var continuation: CheckedContinuation<Void, Never>?

    func activateCallSession(intent: AudioIntent) async throws {}

    func deactivateCallSession() async {
        guard blockNextDeactivation else { return }
        blockNextDeactivation = false
        deactivationInFlight = true
        await withCheckedContinuation { (c: CheckedContinuation<Void, Never>) in
            self.continuation = c
        }
        deactivationInFlight = false
    }

    func releaseDeactivation() {
        let c = continuation
        continuation = nil
        c?.resume()
    }

    func applyRouting(_ device: AudioDevice) async throws {}
    func setMicMuted(_ muted: Bool) async throws {}

    var availableDevices: AsyncStream<[AudioDevice]> { AsyncStream { _ in } }
    var effectiveInputDevice: AsyncStream<AudioDevice?> { AsyncStream { _ in } }
    var effectiveOutputDevice: AsyncStream<AudioDevice?> { AsyncStream { _ in } }
    var events: AsyncStream<AudioCoordinatorEvent> { AsyncStream { _ in } }
}

// MARK: - Auto-joining signaling provider

/// Signaling provider that drives a held room join to completion on its own: on
/// `connect()` it fires `didConnect`, and on `joinRoom()` it fires `didJoin` with
/// the configured participants (deferred through `Task { @MainActor }` to match
/// the real delegate-proxy hop). When `autoJoin == false` it never fires
/// `didJoin`, so the held join times out (used by the failed-room-join test).
final class AutoJoinSignalingProvider: SignalingProvider, @unchecked Sendable {
    let version: Int = SUPPORTED_SIGNALING_PROVIDER_VERSION
    let capabilities = ProviderCapabilities(handlesReconnection: false)
    weak var delegate: SignalingProviderDelegate?

    private let localCid: String
    private let remoteCid: String?
    private let autoJoin: Bool

    init(localCid: String, remoteCid: String?, autoJoin: Bool) {
        self.localCid = localCid
        self.remoteCid = remoteCid
        self.autoJoin = autoJoin
    }

    func connect() {
        let delegate = self.delegate
        Task { @MainActor in
            delegate?.signalingProviderDidConnect(ConnectionInfo(transport: "ws"))
        }
    }

    func disconnect() {}

    func joinRoom(_ roomId: String, options: JoinOptions) {
        guard autoJoin else { return }
        let delegate = self.delegate
        let localCid = self.localCid
        let remoteCid = self.remoteCid
        Task { @MainActor in
            var participants = [SignalingProviderParticipant(peerId: localCid, joinedAt: 1)]
            if let remoteCid {
                participants.append(SignalingProviderParticipant(peerId: remoteCid, joinedAt: 2))
            }
            delegate?.signalingProviderDidJoin(
                JoinedEvent(
                    peerId: localCid,
                    participants: participants,
                    hostPeerId: localCid,
                    maxParticipants: 4
                )
            )
        }
    }

    func leaveRoom() {}
    func endRoom() {}
    func sendToPeer(_ peerId: String, type: String, payload: SignalingPayload?) {}
    func broadcast(type: String, payload: SignalingPayload?) {}
    func getIceServers() async throws -> [IceServerConfig] { [] }
}

// MARK: - Stall-release session stub

/// A `RegistryManagedSession` stub whose `releaseForeground` can be made to NEVER
/// confirm fully-held (`stallRelease = true`), so a test can drive the registry's
/// old-release timeout (Core Invariant 1). The real iOS session confirms held
/// synchronously and so cannot reach that path. Otherwise it behaves like a
/// minimal foreground-capable session: it reports `.inCall`, activates to
/// `.active`, and holds to `.held`/`.inactive`.
@MainActor
final class StallReleaseSession: RegistryManagedSession {
    let roomId: String
    let roomUrl: URL?
    private(set) var mediaRole: CallMediaRole = .held
    private(set) var mediaActivationState: MediaActivationState = .inactive
    var registryMembershipPhase: SerenadaCallPhase = .inCall
    var registryErrorDescription: String? { nil }
    var registryLocalCid: String? { "stub-local" }
    var registryParticipantCount: Int { 2 }
    var registryDesiredAudioEnabled: Bool = true
    var registryDesiredVideoMode: LocalCameraMode? = nil
    var registryActualAudioPublished: Bool { mediaRole == .foreground }
    var registryActualVideoPublished: Bool { false }
    var registryQualitySummary: CallQualitySummary? { nil }

    /// When true, `releaseForeground` does NOT drive the session to fully-held, so
    /// the registry's drain-settle never confirms and its release timeout trips.
    var stallRelease = false
    private(set) var releaseForegroundCalls = 0
    private(set) var token: ForegroundOwnerToken?

    init(roomId: String, roomUrl: URL? = nil) {
        self.roomId = roomId
        self.roomUrl = roomUrl
    }

    func preflightForeground() -> ForegroundPreflight { .ok }
    func missingDesiredForegroundPermissions() -> [MediaCapability] { [] }

    func activateForeground(_ token: ForegroundOwnerToken, generation: Int) throws {
        self.token = token
        mediaRole = .foreground
        mediaActivationState = .active
    }

    func releaseForeground(_ token: ForegroundOwnerToken) {
        releaseForegroundCalls += 1
        guard !stallRelease else {
            // Model a release that has begun but not confirmed fully-held.
            mediaActivationState = .activating
            return
        }
        mediaRole = .held
        mediaActivationState = .inactive
    }

    func awaitForegroundReleaseSettled() async {}

    func abortForegroundActivation(_ token: ForegroundOwnerToken) {
        mediaRole = .held
        mediaActivationState = .inactive
    }

    func registryLeave() {
        mediaRole = .held
        mediaActivationState = .inactive
        registryMembershipPhase = .idle
    }

    func registryEnd() { registryLeave() }
}

// MARK: - Registry harness

/// Builds a ``SerenadaCallRegistry`` whose session factory creates sessions wired
/// with per-call fakes (auto-joining signaling, a recording/blocking coordinator,
/// fake media/audio/clock), recording each session + coordinator so a test can
/// inspect and assert. Each harness owns a DEDICATED `ForegroundMediaArbiter` for
/// isolation (no `.shared` cross-talk between cases).
@MainActor
final class RegistryTestHarness {
    let arbiter: ForegroundMediaArbiter
    let eventLog = EventLog()
    /// The registry runs against this fake clock so timeout tests drive deadlines
    /// via `advance(byMs:)`; normal ops settle in `awaitSettle`'s yield-only phase.
    let fakeRegistryClock = FakeSessionClock()
    private(set) var registry: SerenadaCallRegistry!

    /// Per-room build options, looked up by canonical roomId when the registry's
    /// factory is invoked.
    private struct RoomOptions {
        var coordinator: SerenadaAudioCoordinator?
        var defaultAudioEnabled: Bool
        var grant: (MediaCapability) -> Bool
        var autoJoin: Bool
        /// When set, the factory returns this stub instead of a real session — for
        /// paths the real (synchronous-release) session cannot reach.
        var stub: RegistryManagedSession?
    }
    private var roomOptions: [String: RoomOptions] = [:]
    /// The session created per canonical roomId (so `session(for:)` resolves it).
    private var sessionsByRoom: [String: SerenadaSession] = [:]
    /// The coordinator created per canonical roomId.
    private var coordinatorsByRoom: [String: SerenadaAudioCoordinator] = [:]

    init(arbiter: ForegroundMediaArbiter? = nil) {
        // A caller-supplied dedicated arbiter is used as-is; otherwise mint a fresh
        // one so this harness never touches the process singleton.
        self.arbiter = arbiter ?? ForegroundMediaArbiter()
        let dedicatedArbiter = self.arbiter
        // Build a throwaway core only to satisfy the registry's `core` dependency;
        // the injected factory never calls `core.makeManagedSession`, so the core's
        // own signaling config is irrelevant.
        let core = SerenadaCore(config: SerenadaConfig(serverHost: "serenada.app"))
        self.registry = SerenadaCallRegistry(
            core: core,
            arbiter: dedicatedArbiter,
            clock: fakeRegistryClock,
            sessionFactory: { [weak self] room, role in
                self!.makeSession(room: room, role: role, arbiter: dedicatedArbiter)
            }
        )
    }

    /// A `RoomRef` for a 27-char room token, recording per-room build options.
    func room(
        _ token: String,
        coordinator: SerenadaAudioCoordinator? = nil,
        defaultAudioEnabled: Bool = true,
        grant: @escaping (MediaCapability) -> Bool = { _ in true },
        autoJoin: Bool = true,
        stub: RegistryManagedSession? = nil
    ) -> RoomRef {
        roomOptions[token] = RoomOptions(
            coordinator: coordinator,
            defaultAudioEnabled: defaultAudioEnabled,
            grant: grant,
            autoJoin: autoJoin,
            stub: stub
        )
        return RoomRef(url: URL(string: "https://serenada.app/call/\(token)")!)
    }

    func session(for id: CallId) -> SerenadaSession? {
        registry.call(id: id)?.session
    }

    func roomId(of id: CallId) -> String {
        registry.call(id: id)?.roomId ?? ""
    }

    func coordinator(for id: CallId) -> RecordingAudioCoordinator? {
        guard let roomId = registry.call(id: id)?.roomId else { return nil }
        return coordinatorsByRoom[roomId] as? RecordingAudioCoordinator
    }

    private func makeSession(
        room: RoomRef,
        role: CallMediaRole,
        arbiter: ForegroundMediaArbiter
    ) -> RegistryManagedSession {
        let url = room.url!
        let roomId = DeepLinkParser.extractRoomId(from: url) ?? url.lastPathComponent
        let opts = roomOptions[roomId] ?? RoomOptions(
            coordinator: nil, defaultAudioEnabled: true, grant: { _ in true }, autoJoin: true, stub: nil
        )
        if let stub = opts.stub { return stub }
        let localCid = "local-\(roomId.prefix(4))"
        let remoteCid = opts.autoJoin ? "remote-\(roomId.prefix(4))" : nil
        let provider = AutoJoinSignalingProvider(localCid: localCid, remoteCid: remoteCid, autoJoin: opts.autoJoin)
        let coordinator = opts.coordinator ?? RecordingAudioCoordinator(tag: roomId, log: eventLog)
        coordinatorsByRoom[roomId] = coordinator

        let config = SerenadaConfig(
            signalingProvider: provider,
            defaultAudioEnabled: opts.defaultAudioEnabled,
            defaultVideoEnabled: false,
            audioCoordinator: coordinator
        )
        let session = SerenadaSession(
            roomId: roomId,
            roomUrl: url,
            config: config,
            initialSignalingProvider: provider,
            audioController: FakeAudioController(),
            mediaEngine: FakeMediaEngine(),
            clock: FakeSessionClock(),
            initialMediaRole: role,
            acquireForegroundLease: false,
            isCapabilityGranted: opts.grant,
            foregroundArbiter: arbiter
        )
        sessionsByRoom[roomId] = session
        return session
    }

    func teardown() async {
        // Leave every non-ended call so each session releases resources and the
        // dedicated arbiter ends clean (no leaked lease/mode).
        for state in registry.calls where state.membershipPhase != .idle {
            await registry.leaveCall(id: state.id)
        }
        arbiter.resetForTests()
        ForegroundMediaArbiter.shared.resetForTests()
    }
}
